SNT/TypeGenerator.tt

```diff
@@ -0,0 +1,160 @@
+<#@ template language="C#" #>
+<#@ assembly name="System.Core" #>
+<#@ import namespace="System.Linq" #>
+<#@ import namespace="System.Text" #>
+<#@ import namespace="System.Collections.Generic" #>
+using System;
+using System.Linq;
+using System.Numerics.Tensors;
+using TorchSharp;
+
+namespace Torch.SNT 
+{
+<#
+foreach (var type in new [] { 
+    new { ApiName = "Byte", Storage = "byte", AccReal="long" },
+    new { ApiName = "Short", Storage = "short", AccReal="long" },   
+    new { ApiName = "Int", Storage = "int",  AccReal="long"},
+    new { ApiName = "Long", Storage = "long", AccReal="long" },
+    new { ApiName = "Double", Storage = "double", AccReal="double" },
+    new { ApiName = "Float", Storage = "float", AccReal="double" },
+     }) {
+        string tname = type.ApiName;
+#>
+    /// <summary>
+    ///   Wrapper class used to surface a Torch <#=tname#>Tensor as a System.Numerics DensorTensor of <#=tname.ToLower()#>
+    /// </summary>
+    public sealed class <#=tname#>TorchTensor : DenseTensor<<#=tname.ToLower()#>>
```

Comment:
Can you summarize briefly why you needed to subclass DenseTensor rather than use it as is?  Was it only so that you could make the Clone give you back a tensor with memory owned by torch rather than managed memory?

Reply:
Thank you for this.  Did you consider making your derived type generic as well?

Reply:
I think my initial implementation was generic but we figured that a non-generic version was simpler. What do you think? Do you think that a generic type will be more "compatible" with the `Tensor<T>` design?

Reply:
If you have a generic type / methods it enables generic callers, if you don't have them then that scenario is not possible (unless the caller creates a type-switch).  Ideally we want to handle the type switching in the libraries rather than force all the callers to do type switching.

We also try to design generic APIs so that the generic type can be inferred: this is why we have the static creation methods on Tensor (note: not Tensor<T>) rather than forcing folks to call the constructor.

Reply:
In our case we force users to create tensors by calling the proper static object (e.g., `FloatTorchTensor.Create`). The main problem we had with the generic approach is that since the inner torch tensors are not generic, each function call would have to contain a switch statement. I think until with don't have a proper generic type hierarchy for the inner torch tensor type (issue #18) having a not generic type is the only reasonable possibility. But I agree that we should try to have a generic API.

Reply:
The underlying C APIs are element-type-specific, so without something like C++'s template specialization available to us, how would we implement a generic concrete Tensor class?

Reply:
Like here #28.

Reply:
> The main problem we had with the generic approach is that since the inner torch tensors are not generic, each function call would have to contain a switch statement.

Tensor<T> handles this specialization already in other cases.  Consider the arithmetic functions where we needed a bridge from generic type to type-specific IL in all function calls.  We have a mapping that happens per-managed type and gets stored in a singleton: https://github.com/dotnet/corefx/blob/master/src/System.Numerics.Tensors/tests/TensorArithmetic.cs#L44.  This isn't the most elegant solution but it works and is relatively lightweight.  I believe @tannergooding also had some ideas to improve this.

Reply:
FWIW, libtorch C++ API *is not* generic - that is, it has a single non-parameterized `at::Tensor` class with methods like `.type()`, `.device()`, `.is_sparse()` etc. All dispatching to the proper backend and function specialization happens at runtime in `Type` and `TensorImpl` classes. So maybe if even libtorch developers were not able to make the `Tensor` class parameterized - not even by the scalar type and in less restrictive C++ template system, we should not worry about it too much, either?

P.S. That said, I'd be delighted to see a strongly typed tensor library for C# - ideally something in the spirit of Haskell's [HMatrix](http://dis.um.es/~alberto/hmatrix/hmatrix.html). The question is how much effort we want to put into it. (Maybe if we have a nice low-level API now, we can build an alternative strongly-typed parameterized API on top of it later?) Just thinking out loud.

Reply:
These are two different concerns tho. I think that the `Tensor<T>` wrapper for TorchSharp should be following the `Tensor<T>` paradigm, i.e., try to use generics. For the main TorchSharp story, I totally agree with you: it is fine to be non-generic and consistent with the Torch API, even if a generic API will be better (and Issue #18 is exactly for this).

Reply:
I was looking into following the approach of System.Numerics.Tensors for dispatching types but it is not going to work in our case at the moment. The problem is that we don't have a root type for the tensors in TorchSharp (i.e., the `ITensorArithmetic<T>` in https://github.com/dotnet/corefx/blob/master/src/System.Numerics.Tensors/tests/TensorArithmetic.cs) therefore I am not able to create a singleton object for type dispatching.

Reply:
That's just an internal interface, could't you create one?